
# Module 61.1: Curve-Based and Empirical Fixed-Income Risk Measures

<hr>


In [ ]:
# Import libraries
import math
import pandas as pd
import numpy as np
import numpy_financial as npf
# import matplotlib.pyplot as plt
# import seaborn as sns
# from sklearn.model_selection import train_test_split
# from pandas.core.common import random_state
# from sklearn.linear_model import LinearRegression

### LOS 61.a: Explain why effective duration and effective convexity are the most appropriate measures of interest rate risk for bonds with embedded options.


### LOS 61.b: Calculate the percentage price change of a bond for a specified change in benchmark yield, given the bond's effective duration and convexity.

So far, all of our duration measures have been calculated using the YTM and prices of straight (option-free) bonds. This is straightforward because both the future cash flows and their timing are known. This is not the case with bonds that have embedded options, such as callable bonds, putable bonds, or a mortgage-backed security (MBS). Embedded options may bring about the early termination of a bond, either at the choice of the investor (for a putable bond) or at the choice of the issuer or underlying borrowers (for callable bonds and MBSs).

>**Professor's Note**
\
Recall that a callable bond gives the issuer the right to buy the bond back before maturity. As such, it is equivalent to a straight (option-free) bond and a short call option position. A putable bond gives the investor the right to sell the bond back to the issuer before maturity. As such, it is equivalent to a straight (option-free) bond and a long put option position.
\
MBSs are similar to callable bonds because mortgage borrowers have the right to prepay their loans. We will explain MBSs in detail in our reading on Mortgage-Backed Security Instrument and Market Features.

The fact that bonds with embedded options have uncertain future cash flows and redemption dates means they do not have a single well-defined yield. The yield of the bond will depend on whether the option embedded in the bond is exercised (recall that we can calculate a yield to maturity and a yield to each call date for a callable bond).

Thus, analyzing interest rate risk for bonds with embedded options is based on shifts in the benchmark curve (e.g., government par rates), rather than changes in the bond's own yield. This measure of price sensitivity is referred to as **effective duration** (`EffDur`).

Calculating effective duration is the same as calculating approximate modified duration, but we replace the change in YTM with $\Delta$ Curve, the change in the benchmark yield curve (used with a bond pricing model to generate $V_−$ and $V_+$). The formula for calculating effective duration is as follows:

$$
\text{effective duration} = \frac{V_{−}+V_{+}-2V_{0}}{2V_{0}\Delta \text{Curve}}
$$

Another difference between effective duration and the methods we have discussed so far is that effective duration separates the effects of changes in benchmark yields from changes in the spread for credit and liquidity risk. Modified duration makes no distinction between changes in the benchmark yield and changes in the spread. Effective duration reflects only the sensitivity of the bond's value to changes in the benchmark yield curve and assumes all else (including spreads) remains the same.

When calculating the convexity of bonds with embedded options, we use an analogous measure, **effective convexity** (`EffCon`), which is once again based on changes in the benchmark curve, rather than on changes in the bond's YTM:

$$
\text{effective convexity} = \frac{V_{−}+V_{+}-2V_{0}}{(\Delta \text{Curve})^2V_{0}}
$$


While the convexity of any option-free bond is positive, a callable bond can exhibit **negative convexity**. This is because at low yields, the call option becomes more valuable, and the call price puts an effective limit on increases in bond value, as shown in  **Price-Yield Function of Callable vs. Option-Free Bond With a Call Price of 102**. For a bond with negative convexity, the price increase that results from a decrease in YTM is smaller than the price decrease that results from equal-sized increase in YTM. This means the duration of a callable bond is less than that of an equivalent option-free bond at low yields.

*Price-Yield Function of Callable vs. Option-Free Bond With a Call Price of 102*

<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Fixed%20Income/Module%2061/pics/convex.jpeg?raw=true">

A putable bond always has positive convexity. In  **Comparing the Price-Yield Curves for Option-Free and Putable Bonds**, we illustrate the price-yield relation for a putable bond. At higher yields, the put becomes more valuable, so that the value of the putable bond decreases less than that of an option-free bond as yield increases. This means the duration of a putable bond is less than that of an equivalent option-free bond at high yields.

*Comparing the Price-Yield Curves for Option-Free and Putable Bonds*

<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Fixed%20Income/Module%2061/pics/convex2.jpeg?raw=true">


For an option-free bond, small differences can be observed between modified duration (with respect to $\Delta$ YTM) and effective duration (with respect to $\Delta$ Curve). This may be surprising, because it seems natural to assume that any shift in government par yield curve would automatically flow through into a similar change in the yields of risky bonds. However, in a nonflat yield curve environment, a shift in the benchmark par yield curve will generate a *nonparallel* shift in the government spot curve. If we assume that credit spreads above government spot rates remain the same, this will mean the change in the risky bond's yield will be slightly different to the original shift in government par yield curves, causing `ModDur` to be slightly different to `EffDur`.


>**Professor's Note**
/
Recall that bond yields are a weighted average of the spot rates that apply to the individual cash flows of the bond. When we shift the government par yield curve to calculate `EffDur`, we must also change the government spot rate curve that generates these yields. Not all spot rates have an equal weight in calculating a particular par yield; hence, spot rates change by different amounts (the spot curve undergoes a nonparallel shift). This nonparallel shift in spots means that the yields of risky bonds, assumed to offer a constant spread above government spot rates, may not move exactly in line with the original $\Delta$ Curve shift. This is a fairly technical point, so for the exam just understand that for option-free bonds, `ModDur` and `EffDur` are not exactly the same for a given ∆Curve unless the yield curve is flat.


We can estimate the expected price change for a bond with respect to an expected change in the yield curve using `EffDur` and `EffCon` in the same way we use modified duration and convexity with respect to $\Delta$ YTM:

$$
\textbf{change in full bond price} = \\
\text{-EffDur} × (\Delta \text{Curve}) + \frac{1}{2} × \text{EffCon} × (\Delta \text{Curve})^2 
$$

Unlike modified duration and convexity, effective duration and convexity do not necessarily provide better estimates of bond prices for smaller changes in yield. This is because for bonds with embedded options, considerations other than the level of government rates determine whether the option is likely to be exercised (e.g., the level of credit spreads on a corporate bond or the amount of principal outstanding on a mortgage).


### LOS 61.c: Define key rate duration and describe its use to measure price sensitivity of fixed-income instruments to benchmark yield curve changes.

Recall that effective duration is an adequate measure of bond price risk only for parallel shifts in the benchmark yield curve. The impact of nonparallel shifts can be measured using a concept known as key rate duration. A **key rate duration**, also known as a **partial duration**, is defined as the sensitivity of the value of a bond or portfolio to changes in the benchmark yield for a *specific* maturity, holding other yields constant. The sum of a bond's key rate durations equals its effective duration.

Key rate duration is particularly useful for measuring **shaping risk**, defined as the effect of a nonparallel shift in the yield curve on a bond portfolio. We can use the key rate duration for each maturity to compute the effect on the portfolio of a yield change at that maturity. The effect on the overall portfolio is the sum of these individual effects.

The key rate duration of a cash flow in a portfolio is the cash flow's modified duration multiplied by its weight in the portfolio. This key rate duration can then be used to assess the impact of a nonparallel shift in the yield curve, as demonstrated in the following example.

<hr>

**EXAMPLE: KEY RATE DURATION**

A portfolio has equally weighted investments in a 5-year zero-coupon bond yielding 5% and a 10-year bond yielding 6%. Yields are quoted on an annual coupon basis. What would be the performance of the portfolio if 5-year yields increase by 50 basis points and 10-year yields decrease by 25 basis points?

In [89]:
# For the 5-year cash flow: 
ModDur = 5 / (1.05)
print("For the 5-year cash flow: ", round(ModDur,3))

# weight in portfolio
weight = 0.5

# The 5-year key rate duration is ModDur * weight in portfolio 
five_kr_duration = ModDur * 0.5
print("The 5-year key rate duration is ModDur x weight in portfolio: ",
      round(five_kr_duration,3))

# The impact of a 50 bp increase in the 5-year yield is, therefore,
fiftybpsincr = five_kr_duration * 0.0050
print("The impact of a 50 bp increase in the 5-year yield, therefore:",round(fiftybpsincr,5))
print ("or equally expressed in percentage increased the 5-year yield: +{:.02%}".format(fiftybpsincr)) 

# For the 10-year cash flow: 
ModDur2 = 10 / (1.06)
print("For the 10-year cash flow the modified duration is: ",round(ModDur2,3))

# The 10-year key rate duration is 
ten_kr_duration = ModDur2 * weight
print("The 10-year key rate duration is: ", round(ten_kr_duration,3))

# The impact of a 25 bp decrease in the 10-year yield is, therefore, 
twentyfivebpsincr = ten_kr_duration * 0.0025
print("The impact of a 25 bp decrease in the 10-year yield is, therefore:",round(twentyfivebpsincr,4))
print ("or equally expressed in percentage increased the 10-year yield: +{:.02%}".format(twentyfivebpsincr)) 


# Overall, the portfolio value will change by 
valchange = (fiftybpsincr*-1) + twentyfivebpsincr
# print(round(valchange,3))
print("-------------------------------------------------------------------------")
print ("Overall, the portfolio value will change by {:.02%}".format(valchange)) 
print()
print("The portfolio value is expected to remain roughly unchanged in \n response to this nonparallel shift in yields.")

For the 5-year cash flow:  4.762
The 5-year key rate duration is ModDur x weight in portfolio:  2.381
The impact of a 50 bp increase in the 5-year yield, therefore: 0.0119
or equally expressed in percentage increased the 5-year yield: +1.19%
For the 10-year cash flow the modified duration is:  9.434
The 10-year key rate duration is:  4.717
The impact of a 25 bp decrease in the 10-year yield is, therefore: 0.0118
or equally expressed in percentage increased the 10-year yield: +1.18%
-------------------------------------------------------------------------
Overall, the portfolio value will change by -0.01%

The portfolio value is expected to remain roughly unchanged in 
 response to this nonparallel shift in yields.


**Answer:**

Recall that modified duration is equal to Macaulay duration divided by (1 + periodic yield). Also recall that the Macaulay duration of a single cash flow is equal to its maturity.

For the 5-year cash flow: ModDur = 5 / (1.05) = 4.762

The 5-year key rate duration is ModDur × weight in portfolio = 4.762 × 0.5 = 2.381.

The impact of a 50 bp increase in the 5-year yield is, therefore, –2.381 × 0.0050 = –0.0119, or –1.19%.

For the 10-year cash flow: ModDur = 10 / (1.06) = 9.434

The 10-year key rate duration is ModDur × weight in portfolio = 9.434 × 0.5 = 4.717.

The impact of a 25 bp decrease in the 10-year yield is, therefore, –4.717 × –0.0025 = 0.0118, or 1.18%.

Overall, the portfolio value will change by –1.19% + 1.18% = –0.01%. 

The portfolio value is expected to remain roughly unchanged in response to this nonparallel shift in yields.

### LOS 61.d: Describe the difference between empirical duration and analytical duration.

The duration measures we have introduced so far, based on mathematical analysis, are often referred to as **analytical durations**. A different approach is to estimate **empirical duration**s using the actual observed historical relationship between benchmark yield changes and bond price changes.

When we estimate corporate bond durations based on a shift in the benchmark (government) yield curve, we implicitly assume that the credit spread for the corporate bond remains unchanged (i.e., changes in the benchmark yield curve and a bond's yield spread are uncorrelated). When this assumption is not justified, estimates of empirical duration, based on the actual relationship between changes in the benchmark yield curve and bond values, may be more appropriate.

An example of such a situation is an increase in market uncertainty during which investor demand shifts sharply toward bonds with low credit risk (a "flight to quality"). Yields on government bonds decrease, but credit spreads increase at the same time. As a result, government bond prices increase, but corporate bond prices increase by less or possibly even fall. For a corporate bond portfolio, an estimate of empirical duration that accounts for this effect would be lower (i.e., less price response to a decrease in benchmark yields) than an estimate of analytical duration would indicate. An analytical estimate of the duration of a portfolio consisting primarily of government debt securities, in this case, would still be appropriate, while an empirically derived estimate of duration would be more appropriate for a portfolio comprising corporate bonds (risky credits).



<hr>

## Reading 61: Key Concepts

<hr>

### LOS 61.a

Because bonds with embedded options have uncertain cash flows, they do not have a single well-defined yield. Therefore, effective duration and effective convexity must be calculated with respect to shifts in the benchmark curve rather than the bond's yield for bonds with embedded options.

Effective duration is a linear estimate of the percentage change in a bond's price that would result from a 1% change in the benchmark yield curve:

$$
\text{effective duration} = \frac{V_{−}+V_{+}-2V_{0}}{2V_{0}\Delta \text{Curve}}
$$
\
\
$$
\text{effective convexity} = \frac{V_{−}+V_{+}-2V_{0}}{(\Delta \text{Curve})^2V_{0}}
$$
\
Callable bonds and MBS may exhibit negative convexity at low yields.

### LOS 61.b

The expected price change for a bond with respect to an expected $\Delta$Curve is estimated as follows:

\
$$
\textbf{change in full bond price} = \\
\text{-EffDur} × (\Delta \text{Curve}) + \frac{1}{2} × \text{EffCon} × (\Delta \text{Curve})^2 
$$

### LOS 61.c

Key rate duration is a measure of the price sensitivity of a bond or a bond portfolio to a change in yield for a specific maturity while other yields remain the same. Key rate durations of a bond or portfolio can be used to estimate price sensitivity to changes in the shape of the yield curve.

### LOS 61.d

Macaulay, modified, and effective duration are examples of analytical duration. Empirical duration is estimated from historical data using models. Empirical duration may be lower than analytical duration in interest rate environments where the assumptions underlying analytical duration may not hold, such as for credit-risky bonds in a flight-to-quality scenario.

